haskell/src/Stablecoin/Client/Impl.hs

```diff
@@ -0,0 +1,322 @@
+setTransferlist sender contract transferlist = do
+  transferlistAddr <- traverse resolveAddress transferlist
+  call sender contract (Call @"Set_transferlist") transferlistAddr
+
+-- | Retrieve the contract's balance.
+getBalance :: "contract" :! AddressOrAlias -> MorleyClientM Mutez
+getBalance (arg #contract -> contract) = do
+  contractAddr <- resolveAddress contract
+  Client.getBalance contractAddr
+
+-- | Check whether the contract has been paused.
+isPaused :: "contract" :! AddressOrAlias -> MorleyClientM Bool
+isPaused contract =
+  getStorage contract <&> \case
+    ((_, (_, arg #paused -> paused)), _) -> paused
+
+-- | Get the address and optional alias of the current contract owner.
+getContractOwner :: "contract" :! AddressOrAlias -> MorleyClientM AddressAndAlias
+getContractOwner contract =
+  getStorage contract >>= \case
+    StorageViewRoles (OwnerRole ownerAddr) -> pairWithAlias ownerAddr
+
+-- | Check if there's an ownership transfer pending acceptance and, if so,
+-- get its pending owner's address and optional alias.
+getPendingContractOwner :: "contract" :! AddressOrAlias -> MorleyClientM (Maybe AddressAndAlias)
+getPendingContractOwner contract =
+  getStorage contract >>= \case
+    StorageViewRoles (PendingOwnerRole ownerAddr) -> traverse pairWithAlias ownerAddr
+
+-- | Get the address and optional alias of the master minter.
+getMasterMinter :: "contract" :! AddressOrAlias -> MorleyClientM AddressAndAlias
+getMasterMinter contract =
+  getStorage contract >>= \case
+    StorageViewRoles (MasterMinterRole masterMinterAddr) -> pairWithAlias masterMinterAddr
+
+-- | Get the address and optional alias of the pauser.
+getPauser :: "contract" :! AddressOrAlias -> MorleyClientM AddressAndAlias
+getPauser contract =
+  getStorage contract >>= \case
+    StorageViewRoles (PauserRole pauserAddr) -> pairWithAlias pauserAddr
+
+-- | Check if a transferlist is set and, if so, get its address and optional alias.
+getTransferlist :: "contract" :! AddressOrAlias -> MorleyClientM (Maybe AddressAndAlias)
+getTransferlist contract =
+  getStorage contract >>= \case
+    (_ , ((_, arg #transferlist_contract -> transferlistMb), _)) -> traverse pairWithAlias transferlistMb
+
+-- | Get the total supply of tokens.
+getTotalSupply :: "contract" :! AddressOrAlias -> MorleyClientM Natural
+getTotalSupply contract =
+  getStorage contract <&> \case
+    (_, (_, (_, arg #total_supply -> totalSupply))) -> totalSupply
+
+-- | Check if there's an alias associated with this address and, if so, return both.
+pairWithAlias :: Address -> MorleyClientM AddressAndAlias
+pairWithAlias addr = do
+  aliasMb <- (Just <$> getAlias (AddressResolved addr)) `catch` \case
+    UnknownAddress _ -> pure Nothing
+    err -> throwM err
+  pure $ AddressAndAlias addr aliasMb
+
+-- | Call an entrypoint of the stablecoin contract with the given argument.
+call
+  :: forall epRef epArg.
+     HasEntryPointArg Parameter epRef epArg
+  => IsoValue epArg
+  => Typeable epArg
```

Comment:
It looks somewhat unusual to write constraints as `a => b => c` instead of `(a, b) => c`. Do you prefer this style or does it have any advantages?

Reply:
No particular advantages, though I personally like the similarity to `a -> b -> c` :) Should I change it for consistency?